grpc/src/jmh/java/com/linecorp/armeria/grpc/shared/SimpleBenchmarkBase.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.grpc.shared;
+
+import static com.linecorp.armeria.grpc.shared.GithubApiService.SEARCH_RESPONSE;
+import static org.awaitility.Awaitility.await;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.annotation.Nullable;
+
+import org.openjdk.jmh.annotations.AuxCounters;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.Empty;
+
+import com.linecorp.armeria.benchmarks.GithubApi.SearchResponse;
+import com.linecorp.armeria.benchmarks.GithubServiceGrpc;
+import com.linecorp.armeria.benchmarks.GithubServiceGrpc.GithubServiceBlockingStub;
+import com.linecorp.armeria.benchmarks.GithubServiceGrpc.GithubServiceFutureStub;
+
+import io.grpc.ManagedChannel;
+import io.grpc.okhttp.OkHttpChannelBuilder;
+
+@State(Scope.Benchmark)
+public abstract class SimpleBenchmarkBase {
+
+    @AuxCounters
+    @State(Scope.Thread)
+    public static class Counters {
+        private AtomicLong numRequests = new AtomicLong();
+        private AtomicLong numFailures = new AtomicLong();
+        private AtomicLong currentRequests = new AtomicLong();
+
+        private volatile boolean waiting;
+
+        public long numRequests() {
+            return numRequests.get();
+        }
+
+        public long numFailures() {
+            return numFailures.get();
+        }
+
+        public long currentRequests() {
+            return currentRequests.get();
+        }
+
+        @Setup(Level.Iteration)
+        public void reset() {
+            waiting = false;
+            numRequests.set(0);
+            numFailures.set(0);
+            currentRequests.set(0);
+        }
+
+        @TearDown(Level.Iteration)
+        public void waitForCurrentRequests() {
+            waiting = true;
+            await().forever().until(() -> currentRequests.get() == 0);
```

Comment:
Question: does JMH take the changes during iteration teardown into account?

Reply:
JMH returns the value of the counters at the end of teardown, so while it doesn't add the time of the teardown to the run, the counters would be affected. It's why I added the `waiting` guard.